### PR TITLE
[IMP] hr_holidays: out of office for global time off

### PR DESCRIPTION
**Pull request by jayp-odoo**
**Public holidays** are now treated like regular out-of-office status in Discuss.
and the holiday is computed on hr.employee from the employee’s company calendar
(only when they have no validated leave). The Discuss banner,
IM status, and formatted display name show
“back on XXX” when appropriate,

task-2377854

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_holidays/static/src/res_partner_model_patch.js [hunk 1]
```diff
@@ -21,6 +21,9 @@ patch(ResPartner.prototype, {
     get outOfOfficeDateEndText() {
         const employee_id = this.employee_id || this.main_user_id?.employee_id;
         if (!employee_id?.leave_date_to) {
+            if (this.main_user_id?.on_public_leave && this != this.store.odoobot) {
```
Comment (rapr-odoo) [inline addons/hr_holidays/static/src/res_partner_model_patch.js]: There should be a better way to do the `odoobot` check

File: addons/hr_holidays/static/src/res_partner_model_patch.js [hunk 2]
```diff
@@ -21,6 +21,9 @@ patch(ResPartner.prototype, {
...
+                return _t("Out of office due to public holiday");
+            }
             return "";
```
Comment (rapr-odoo) [inline addons/hr_holidays/static/src/res_partner_model_patch.js]: Could be inlined

File: addons/hr_holidays/models/res_users.py [hunk 3]
```diff
@@ -10,6 +10,25 @@ class ResUsers(models.Model):
     _inherit = "res.users"
 
     leave_date_to = field_employee(fields.Date, 'leave_date_to')
+    on_public_leave = fields.Boolean(compute="_compute_on_public_leave")
+
+    def _compute_on_public_leave(self):
```
Comment (rapr-odoo) [inline addons/hr_holidays/models/res_users.py]: Shouldn't we add some dependency for this compute?

File: addons/hr_holidays/models/res_users.py [hunk 4]
```diff
@@ -10,6 +10,25 @@ class ResUsers(models.Model):
...
+        now = fields.Datetime.now()
+        company_ids = self.mapped("company_id").ids
```
Comment (rapr-odoo) [inline addons/hr_holidays/models/res_users.py]: Would this work in our production?

File: addons/hr_holidays/models/res_users.py [hunk 5]
```diff
@@ -31,13 +50,28 @@ def _get_on_leave_ids(self, partner=False):
...
+                EXISTS (
+                    SELECT 1 FROM hr_leave
+                    WHERE hr_leave.user_id = res_users.id
+                    AND hr_leave.state = 'validate'
+                    AND hr_leave.date_from <= %s
+                    AND hr_leave.date_to >= %s
+                )
+                OR EXISTS (
+                    SELECT 1 FROM resource_calendar_leaves rcl
+                    WHERE rcl.resource_id IS NULL
```
Comment (rapr-odoo) [inline addons/hr_holidays/models/res_users.py]: What's the need for the check on `resource_id`?

File: addons/hr_holidays/models/res_users.py [hunk 6]
```diff
@@ -31,13 +50,28 @@ def _get_on_leave_ids(self, partner=False):
         field = 'partner_id' if partner else 'id'
         self.flush_model(['active'])
         self.env['hr.leave'].flush_model(['user_id', 'state', 'date_from', 'date_to'])
-        self.env.cr.execute('''SELECT res_users.%s FROM res_users
-                            JOIN hr_leave ON hr_leave.user_id = res_users.id
-                            AND hr_leave.state = 'validate'
-                            AND res_users.active IS TRUE
-                            AND hr_leave.date_from <= %%s AND hr_leave.date_to >= %%s
-                            RIGHT JOIN hr_leave_type ON hr_leave.holiday_status_id = hr_leave_type.id
```
Comment (rapr-odoo) [inline addons/hr_holidays/models/res_users.py]: Why is this check removed in new query?

File: addons/hr_holidays/models/res_users.py [hunk 7]
```diff
@@ -64,16 +98,23 @@ def create(self, vals_list):
...
+                name = "%s \t ✈ --%s %s--" % (user.display_name or user.name, _("Back on"), back_on)
+                user.display_name = name.strip()
+            elif user.on_public_leave:
+                name = "%s \t ✈ --%s--" % (user.display_name or user.name, _("Out of office due to public holiday"))
                 user.display_name = name.strip()
 
     def _store_main_user_fields(self, res: Store.FieldList):
         super()._store_main_user_fields(res)
         if res.is_for_internal_users():
+            res.attr("on_public_leave", predicate=lambda user: user.on_public_leave)
```
Comment (rapr-odoo) [inline addons/hr_holidays/models/res_users.py]: Don't we need to add this in the res model in JS?

File: addons/hr_holidays/models/res_users.py (hunk 5 above)
Comment (jayp-odoo) [reply addons/hr_holidays/models/res_users.py]: when `resource_id` is set, it targets a specific resource/employee and shouldn’t mark everyone as on a public leave. so we need resource_id to be NULL.

File: addons/hr_holidays/models/res_users.py [hunk 8]
```diff
@@ -10,6 +10,28 @@ class ResUsers(models.Model):
     _inherit = "res.users"
 
     leave_date_to = fields.Date(**related_employee_field('leave_date_to'))
+    on_public_leave = fields.Boolean(compute="_compute_on_public_leave")
```
Comment (Atovange) [inline addons/hr_holidays/models/res_users.py]: I don't think this field should be on `ResUsers`, as leaves have to do with employees so `HrEmployee` would be a more appropriate place.

It also causes an inconsistency: if a user is an employee in company A, but has access to company B, they would result on leave only if company B is on public holiday.

This opens the question: when a user is an employee in multiple companies, some on public holiday and some not, we gotta decide whether to show them on leave or not.

Comment (jayp-odoo) [reply addons/hr_holidays/models/res_users.py]: @Atovange Hy thanks for the comment :blush: 

 i tested it on my runbot and its not happening like you said 
> if a user is an employee in company A, but has access to company B, they would result on leave only if company B is on public holiday.

could you please cross check it again, thank you in advance!

Comment (Atovange) [reply addons/hr_holidays/models/res_users.py]: Yes, the exact setup is:
image
image
image
image

**Comment by jayp-odoo** [comment]
@Atovange i have updated the code i think the issues has been resolve and also single and double quotes and technical's  things can be debatable since its HR module.

**Comment by Atovange** [comment]
Thanks for the work Jay :)

I wonder if we shouldn't just collapse the public holiday check inside the `leave_date_to` computation itself. Do we specifically want to give users the information about the user being on public holiday?

I see that public leaves are already taken into account in the compute for `leave_date_to`, but only when the user has an `hr.leave` that "touches" the public holiday (at the beginning, end, or in the middle).

Showing "On Public Holiday" would also hide when the employee would be back if they have leaves planned around the public holiday (which is quite common).
For instance if an employee is on leave the three days after a public holiday, on the day of said holiday we would simply show "On Public Holiday".

Perhaps we should check with the PO to see what's the behavior we want :)

File: addons/hr_holidays/models/hr_employee.py [hunk 9]
```diff
@@ -214,13 +214,41 @@ def _compute_leave_status(self):
...
+                [
+                    ('resource_id', '=', False),
+                    ('company_id', 'in', employees_not_on_leave.company_id.ids),
+                    ('date_from', '<=', now),
+                    ('date_to', '>=', now),
+                ],
+                ['company_id'],
+                ['date_to:max'],
+            )
+            company_public_leave_end = {company.id: date_to for company, date_to in grouped_companies}
```
Comment (Atovange) [inline addons/hr_holidays/models/hr_employee.py]: You can wrap the _read_group in `dict` to save this line.

Also the name is a bit opaque. When it's dictionaries I like being explicit, so in this case something like `leave_end_by_company_id`.
Same with the one below `leave_end_by_employee_id`.

File: addons/hr_holidays/models/hr_employee.py [hunk 10]
```diff
@@ -214,13 +214,41 @@ def _compute_leave_status(self):
...
             'leave_date_from': False,
-            'leave_date_to': False,
             'current_leave_state': False,
             'is_absent': False,
         })
+        employees_on_public_leave = self.browse(employee_public_leave_end)
+        (employees_not_on_leave - employees_on_public_leave).leave_date_to = False
+
+        if employee_public_leave_end:
+            employee_back_on_public = employees_on_public_leave._get_first_working_interval_batch(employee_public_leave_end)
```
Comment (Atovange) [inline addons/hr_holidays/models/hr_employee.py]: Same here with the name, this one is a dict too.

File: addons/hr_holidays/models/hr_employee.py [hunk 11]
```diff
@@ -214,13 +214,41 @@ def _compute_leave_status(self):
...
+            for employee in employees_on_public_leave:
+                back_on = employee_back_on_public.get(employee.id, employee_public_leave_end[employee.id])
+                employee.leave_date_to = back_on.date()
```
Comment (Atovange) [inline addons/hr_holidays/models/hr_employee.py]: This logic feels very similar to the one above (the regular leave_date_to assignment). Couldn't we collect everything in one dictionary and loop through the employees once?

**Comment by Atovange** [comment]
LGTM functionally and the logic too, good work!

File: addons/hr_holidays/models/hr_employee.py [hunk 12]
```diff
@@ -193,34 +193,67 @@ def collect_employees(work_intervals):
...
-            for employee, holiday in employee_holidays.items()
-        })
-
-        for employee, emp_holidays in employee_holidays.items():
-            latest_emp_holiday = emp_holidays[0]
-            employee.leave_date_from = min(emp_holidays.mapped('date_from')).date()
-            employee.leave_date_to = employee_back_on.get(employee.id, latest_emp_holiday.date_to).date()
-            employee.current_leave_state = latest_emp_holiday.state
-            employee.is_absent = any(e_h.work_entry_type_id.count_as == 'absence' for e_h in emp_holidays)
+        holidays_by_employee_id = holidays.grouped(lambda leave: leave.employee_id.id)
```
Comment (alexkuhn) [inline addons/hr_holidays/models/hr_employee.py:207]: Would take opportunity to rename `holidays` by `leaves`, as holiday is a type of leave but not all leaves are holidays. The change of wording helps readability.

Too bad `hr.leave` represents an explicit leave so holidays are not taking into account, but that's more a modeling subtlety. Conceptually an absence from holidays is still a leave.

File: addons/hr_holidays/models/hr_employee.py [hunk 13]
```diff
@@ -193,34 +193,67 @@ def collect_employees(work_intervals):
...
+        back_on_by_employee_id = employees_with_leave_end._get_first_working_interval_batch(leave_end_by_employee_id)
+        for employee in employees_with_leave_end:
+            back_on = back_on_by_employee_id.get(employee.id, leave_end_by_employee_id[employee.id])
+            employee.leave_date_to = back_on.date()
+            emp_holidays = holidays_by_employee_id.get(employee.id)
+            if not emp_holidays:  # the employee is off because of a public holiday
+                continue
+            employee.leave_date_from = min(emp_holidays.mapped('date_from')).date()
+            employee.current_leave_state = emp_holidays[0].state
+            employee.is_absent = any(e_h.work_entry_type_id.count_as == 'absence' for e_h in emp_holidays)
```
Comment (alexkuhn) [inline addons/hr_holidays/models/hr_employee.py:259]: Code would be simpler if we don't mix leaves and holidays.

leaving code of hr.leaves alone above and additional diff for holidays looks better for me: https://github.com/odoo/odoo/pull/285012/changes/85e3660dfd42adb5cdf5b76de85f7642f2618e9c

**Review by alexkuhn — APPROVED** [review]
LGTM otherwise

File: addons/hr_holidays/models/hr_employee.py (hunk 13 above)
Comment (jayp-odoo) [reply addons/hr_holidays/models/hr_employee.py:259]: Agreed, but GITO suggested a different approach, so I went with that. Thanks for the review! I’ve updated it now.

Comment (Atovange) [reply addons/hr_holidays/models/hr_employee.py:259]: @alexkuhn The diff does look simpler in the mentioned version but IMO the method body is harder to read in its entirety.

On top of that there's an issue of performance.
In said version the method `_get_first_working_interval_batch` is called twice and it's a quite heavy computation (involves many calls to various big tables).

We should keep in mind that this compute is also called on computed very often, like for the user `display_name` and IM status.

Comment (Atovange) [reply addons/hr_holidays/models/hr_employee.py:259]: @alexkuhn @jayp-odoo ran a quick benchmark

When batch computing for 50 employees (like on a list view) of one company:

|   | Old version (one pass, mixing) | New version (two passes, separate)
| ------------- | ------------- | ------------- |
| No public holidays  | 21 queries  | 22 queries
| With public holidays  | 51 queries  | 66 queries

And the difference gets more stark as we add more employees and more companies (like 270 employees on 4 companies with public holidays: we save 25 queries).

Comment (alexkuhn) [reply addons/hr_holidays/models/hr_employee.py:259]: Ok sounds good with one pass compared to two passes.
The bump of queries from no public holidays to with public holidays is significant. We should try to reduce it if possible.

While this big amount of queries is only when the employee is not in an ongoing encoded hr.leave, keep in mind that functionally a lot of employees will be in the same situation of holidays branching at once, e.g. a Belgian public holiday, so this can put a significant load for very specific days in the calendar with DBs with lots of employees.

Comment (jayp-odoo) [reply addons/hr_holidays/models/hr_employee.py:259]: In general, I feel the older code with Gito’s suggestions looks better. It does make the diff a bit larger, but I think it’s still worth it considering the proofs. Let me know what you’d like me to implement, and I’ll get it done.

Comment (alexkuhn) [reply addons/hr_holidays/models/hr_employee.py:259]: Yes I agree with diffing more as suggested by GITO in the end ;)

Comment (Atovange) [reply addons/hr_holidays/models/hr_employee.py:259]: @alexkuhn about the cost of the feature itself:

The main cost in terms of queries is the "look-ahead" logic of `_get_first_working_interval_batch`.

It costs ~18 queries to run. So the benchmark looks something like this:

(calling SEPARATE the two passes version and MIXED the one pass version)

| batch | master | SEPARATE | MIXED | look-ahead passes<br>M/S/X |
| :--- | :---: | :---: | :---: | :---: |
| 5 on approved time off | 20 | 20 | 20 | 1/1/1 |
| 20 on the public holiday | 1 | 19 | 19  | 0/1/1 |
| 5 on time off + 20 on the holiday | 20 | 38 | 22 | 1/2/1 |
| 1 on time off + 20 on the holiday | 20 | 38 | 22 | 1/2/1 |

So in short:
- When not in a public holiday it costs only one extra query (checking the calendar leaves).
- When on a public holiday it would now trigger a look-ahead on every compute of every employee (in that company), same as the current logic in master if one employee in the batch is on leave today.

**Comment by jayp-odoo** [comment]
@alexkuhn comments are solved in both PR :blush: & again thanks for the good suggestions @Atovange

**Review by alexkuhn — APPROVED** [review]
@robodoo delegate+ r+
